hazelcast/src/main/java/com/hazelcast/ringbuffer/Ringbuffer.java

```diff
@@ -0,0 +1,259 @@
+     *
+     * @return the sequence of the tail.
+     */
+    long tailSequence();
+
+    /**
+     * Returns the sequence of the head. The head is the side of the ringbuffer where the oldest items in the
+     * ringbuffer are found.
+     *
+     * If the RingBuffer is empty, the head will be one more than the tail.
+     *
+     * The initial value of the head is 0 (1 more than tail).
+     *
+     * @return the sequence of the head.
+     */
+    long headSequence();
+
+    /**
+     * Returns the remaining capacity of the ringbuffer.
+     *
+     * The returned value could be stale as soon as it is returned.
+     *
+     * If retention is not set, the remaining capacity will always be the capacity.
+     *
+     * @return the remaining capacity.
+     */
+    long remainingCapacity();
+
+    /**
+     * Adds an item to the tail of the Ringbuffer. Of there is no space in the Ringbuffer, the add will overwrite the oldest
+     * item in the ringbuffer no matter what the retention policy is. For more control on this behavior, check the
+     * {@link #addAsync(Object, OverflowPolicy)} and the {@link OverflowPolicy}.
+     *
+     * The returned value is the sequence of the added item. Using this sequence you can read the added item.
+     *
+     * <h3>Using the sequence as id</h3>
+     * This sequence will always be unique for this Ringbuffer instance so it can be used as a unique id generator if you are
+     * publishing items on this Ringbuffer. However you need to take care of correctly determining an initial id when any node
+     * uses the ringbuffer for the first time. The most reliable way to do that is to write a dummy item into the ringbuffer and
+     * use the returned sequence as initial id. On the reading side, this dummy item should be discard. Please keep in mind that
+     * this id is not the sequence of the item you are about to publish but from a previously published item. So it can't be used
+     * to find that item.
+     *
+     * @param item the item to add.
+     * @return the sequence of the added item.
+     * @throws java.lang.NullPointerException if item is null.
+     * @see #addAsync(Object, OverflowPolicy)
+     */
+    long add(E item);
+
+    /**
+     * Asynchronously writes an item with a configurable backoff policy.
+     *
+     * If there is space in the ringbuffer, the call will return the sequence of the written item.
+     *
+     * If there is no space, it depends on the overflow policy what happens:
+     * <ol>
+     * <li>{@link OverflowPolicy#OVERWRITE}: we just overwrite the oldest item in the ringbuffer and we violate
+     * the retention</li>
+     * <li>{@link OverflowPolicy#FAIL}: we return -1 </li>
+     * </ol>
+     *
+     * The reason that FAIL exist is to give the opportunity to obey the retention policy. A blocking call could have been
+     * created, but this would trap the caller thread and this is not always desirable. If blocking behavior is required,
```

Comment:
I don't quite understand this part. It's async API so what thread would have been trapped? 

```java
ICompletableFuture<Long> f = ringbuffer.addAsync(item, WAIT);
f.andThen(myHandler);
myThreadIsNotTrapped();
```
?